### ISSUE TITLE
Chapter 3: tricoder error in some cases

### ISSUE THREAD

**Issue by balroggg** [skerkour/black-hat-rust#40]
Try to scan ports of `twitch.tv`(or `ya.ru`/`google.com`)

Error:
```
cargo run --release -- twitch.tv
    Finished release [optimized] target(s) in 0.20s
     Running `target/release/tricoder twitch.tv`
thread 'main' panicked at 'port scanner: Creating socket address: Os { code: 16, kind: ResourceBusy, message: "Device or resource busy" }', src/ports.rs:49:10
```

Maybe this is related to https://github.com/rust-lang/rust/issues/47955
because I have same error(`EMFILE`). Increase open file limit is not help me.

Any thoughts?

**Comment by apertureless** [comment]
I am in chapter 2 but I am also receiving some errors:

```
thread 'thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: Custom { kind: Uncategorized, error: "failed to lookup address information: nodename nor servname provided, or not known" }', src/ports.rs:22:10
```

However only if the threads are set to 256. If I set the threads lower to 100, it seems to work.

**Comment by skerkour** [comment]
Hello,
Sorry for the delay.

It indeed seems to be an issue with the concurrency level and the OSes.

Could you try with the patch proposed in #43 (branch: [fix-dns-40](https://github.com/skerkour/black-hat-rust/tree/fix-dns-40)) where `to_socket_addrs` is called only once?

Thanks!

**Comment by apertureless** [comment]
Yep, this seems to fix the crash for me.
But, my scan results are very inconsistent currently. Need to do some more testing.

**Comment by skerkour** [comment]
Hmmmm

What if you increase the timeout in the `scan_port` function?

**Comment by balroggg** [comment]
I tried fix-dns branch, but crash still exist. For me maybe it's old glib on Arch Linux, for some reason it's outdated, version 2.33-5.

**Comment by skerkour** [comment]
Hi @balroggg thanks for the feedback!

What if you decrease `ports_concurrency` and `subdomains_concurrency` here: https://github.com/skerkour/black-hat-rust/blob/main/ch_03/tricoder/src/main.rs#L29?

**Comment by balroggg** [comment]
Those values works for me, but I tried so many times :smiling_face_with_tear: 
```diff
-    let http_timeout = Duration::from_secs(10);
+    let http_timeout = Duration::from_secs(30);
     let http_client = Client::builder().timeout(http_timeout).build()?;
 
-    let ports_concurrency = 200;
-    let subdomains_concurrency = 100;
+    let ports_concurrency = 40;
+    let subdomains_concurrency = 20;
```

**Comment by apertureless** [comment]
Well I've tried those, but at least for me on OSX 11.6. It is kinda weird, because if I scan the same domain, once the scan takes quite long but returns proper open ports. The other time I scan it, it finishes in half the time, but also not showing all open ports.

But I totally blame shitty macos for that. 

I have 

```rs
// main.rs
let http_timeout = Duration::from_secs(30);
let concurrent_ports = 40;
let concurrent_subdomains = 20;

// portscan.rs

let timeout = Duration::from_secs(3);
```

And on my mac it took > 130s to scan on the tries where proper results came back and > 60s on the scans where not all open ports came back.

I've tested the same code on my ubuntu server and it finished within 11s. Tested multiple times with constant results.
So I would blame mac os for that.
